src/librustc_typeck/check/mod.rs

```diff
@@ -1763,47 +1759,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
             debug!("select_all_obligations_and_apply_defaults: defaults={:?}", default_map);
 
-            // We loop over the unsolved variables, resolving them and if they are
-            // and unconstrainted numberic type we add them to the set of unbound
-            // variables. We do this so we only apply literal fallback to type
-            // variables without defaults.
+            // Examine all unsolved variables, and narrow them to the set that have applicable
+            // defaults. We want to process any unsolved variables that have either an explicit
+            // user default, literal fallback, or are diverging.
             for ty in &unsolved_variables {
                 let resolved = self.infcx().resolve_type_vars_if_possible(ty);
-                if self.infcx().type_var_diverges(resolved) {
-                    demand::eqtype(self, codemap::DUMMY_SP, *ty, self.tcx().mk_nil());
-                } else {
-                    match self.infcx().type_is_unconstrained_numeric(resolved) {
-                        UnconstrainedInt | UnconstrainedFloat => {
-                            unbound_tyvars.insert(resolved);
-                        },
-                        Neither => {}
-                    }
-                }
-            }
-
-            // We now remove any numeric types that also have defaults, and instead insert
-            // the type variable with a defined fallback.
-            for ty in &unsolved_variables {
                 if let Some(_default) = default_map.get(ty) {
```

Comment:
Nit: I personally prefer `is_some()` or perhaps `Some(_)` if your intention is just to check whether it is some.